### PR TITLE
build: set up s390x cross-compilation toolchain

### PR DESCRIPTION
**Pull request by ruixin-bao**
This patch modifies the cockroachdb/builder docker image and
builder/mkrelease.sh script to allow cross compiling s390x binary on an intel machine.

With s390x binary being able to be built, I am wondering if it possible for s390x to join the CI (or at least part of it) so the cross-compiled binary can be tested on a s390x machine? The CI will also be helpful to let us know the test status soon enough so we can quickly prepare a fix for them. Additionally, with a s390x machine, adding docker multi-architecture image is possible as the deployment image can then be built on s390x.

Hardware/VM access can be provided. Any help or advice will be appreciated. 

Thanks!

### CODE REVIEW

**Comment by ruixin-bao** [comment]
@knz 
Hello, we discussed a bit about this PR in https://github.com/cockroachdb/cockroach/issues/55307, wondering if you mind taking a look when you get some cycles? I also have a question about possibility of joining the CI in the PR message and am wondering if you can take a look at that too? I understand you mentioned in the issue that s390x would  likley be the same level of ARM support. I was curious about the CI but did not find too much info about whether ARM has joined CI or not, so I figured to ask in this PR as well. 

Thank you.

**Comment by knz** [comment]
This looks fine to me. @otan can you have a look too?

>  also have a question about possibility of joining the CI in the PR message and am wondering if you can take a look at that too?

I don't understand the question. Can you rephrase it differently?

(FYI we do not build ARM in CI.)

**Comment by ruixin-bao** [comment]
Hi @knz,

> This looks fine to me. @otan can you have a look too?

Thank you for taking a look. 

> I don't understand the question. Can you rephrase it differently?

Sorry for the confusion. I was referring to the question I mentioned in the PR message about CI. Quoting it here just in case:

> With s390x binary being able to be built, I am wondering if it possible for s390x to join the CI (or at least part of it) so the cross-compiled binary can be tested on a s390x machine? The CI will also be helpful to let us know the test status soon enough so we can quickly prepare a fix for them. Additionally, with a s390x machine, adding docker multi-architecture image is possible as the deployment image can then be built on s390x.
Hardware/VM access can be provided. Any help or advice will be appreciated.

Our ultimate goal is hoping to have a s390x binary released and available, it can be a non-official binary and/or you can mention it is built by community. Having a binary would be useful so some of the Z (s390x) users can download it directly. Wondering if that would be doable and what needs to be in place in order to accomplish this? I am more than happy to provide any PRs, testing if needed. Thanks.

**Comment by otan** [comment]
sorry for the delay
letting @jlinder handle this one as his team will be responsible for the builds going forwards.

**Comment by knz** [comment]
FYI today we do not run CI for the ARM targets. It does not even run on the "more supported" macOS (darwin) and Windows target. The status of these various targets is checked as follows:

- we check that the build produces an executable (but do not try to run it) for macOS and windows, using cross-compile inside docker. These executables are linked from our docs site with a "use at your own risk" disclaimer.

- cockroachdb developers and product manager routinely use the macOS executables for internal testing, so we get a little bit of experience like that.

- some of our users/customers use the windows binary, and we get some experience reports from them (but we do not guarantee fast reactions upon bug reports)

- we rely on our community to run asynchronous builds and test runs for ARM.

Coming back to your question:

>  I am wondering if it possible for s390x to join the CI (or at least part of it) so the cross-compiled binary can be tested on a s390x machine?

I think we could add s390x alongside darwin and windows in the list of executables produced upon CI, that could seem reasonable? Unsure. I would say, if we do that then we'd likely want to promote the ARM target to the same status as well.


For automated testing, that is a different problem. As for ARM, macOS and windows, the issue we've had with CI integration in the past is threefold:

- we'd need to populate a pool of teamcity agents for these targets, which would make our pool of agents non-homogeneous. This in turn makes our reasoning about teamcity license billing a bit more complicated.

- we need to implement the CI (TC) configuration and scripts for multi-platform testing. We haven't built experience doing that yet.

- it increases the workload upon every CI run, which makes our CI costs higher. We'd need some management sign-off on that.

The bottom line is:

- I'll defer to @jlinder about whether we could add ARM and s390x alongside macOS and windows in the list of executable builds.
- fully-fledged automated testing in CI for the non-linux targets will likely need to wait a bit more.

**Comment by ruixin-bao** [comment]
Hi @knz,

Thank you for providing such an in-depth explanation on both target status and CI information. Much appreciated.

> The status of these various targets is checked as follows:
  we check that the build produces an executable (but do not try to run it) for macOS and windows, using cross-compile inside docker. These executables are linked from our docs site with a "use at your own risk" disclaimer.
  cockroachdb developers and product managers routinely use the macOS executables for internal testing, so we get a little bit of experience like that.
   some of our users/customers use the windows binary, and we get some experience reports from them (but we do not guarantee fast reactions upon bug reports)
  we rely on our community to run asynchronous builds and test runs for ARM.

I see, I think s390x can follow a similar check process as ARM.

> I think we could add s390x alongside darwin and windows in the list of executables produced upon CI, that could seem reasonable?

That would be great, yes we would love to see this happen.

> For automated testing, that is a different problem. As for ARM, macOS and windows, the issue we've had with CI integration in the past is threefold:

Thank you for being transparent and explaining the issues : ).

> The bottom line is:
I'll defer to @jlinder about whether we could add ARM and s390x alongside macOS and windows in the list of executable builds.
> fully-fledged automated testing in CI for the non-linux targets will likely need to wait a bit more.

Thanks for the consideration of s390x binary. Yes, we understand now CI will likely need to wait a bit more.

**Comment by knz** [comment]
@jlinder for this PR there are two action items, one "lightweight" and one with more work. Maybe we can stagger them:

1. review the PR as-is, which simply adds a new build target. Since we wouldn't (initially) auto-build this, we just need to check that the structure of the change follows what is required for the other targets (e.g. ARM)

2. evaluate the question of whether to auto-build ARM and s390x binaries alongside macOS/windows in the "build executable" CI phase. I do not believe we need to fully answer this right away (i.e. we could merge this PR then evaluate the question afterwards).

**Comment by ruixin-bao** [comment]
gentle ping, wondering if there are any updates for the above mentioned action items? Thanks.

**Comment by otan** [comment]
+ @rickystewart since @jlinder is away

**Review by rickystewart — APPROVED** [review]
I have no problem with this -- AFAICT we can take this change while we evaluate what additional work, if any, makes sense for CI.

**Comment by ruixin-bao** [comment]
Thank you for the reviews and consideration.
